Source/WebKit/UIProcess/API/gtk/WebKitPrintOperation.cpp

```diff
@@ -379,20 +415,268 @@ static void webkitPrintOperationPrintPagesForFrame(WebKitPrintOperation* printOp
+
+    auto* connection = g_dbus_proxy_get_connection(priv->portalProxy.get());
+    auto uniqueName = String::fromUTF8(g_dbus_connection_get_unique_name(connection));
+    auto senderName = makeStringByReplacingAll(uniqueName.substring(1), '.', '_');
+    auto token = makeString("WebKitGTK", weakRandomNumber<uint32_t>());
+    auto requestPath = makeString("/org/freedesktop/portal/desktop/request/", senderName, "/", token);
+
+    RELEASE_ASSERT(!priv->signalId);
+
+    priv->signalId = g_dbus_connection_signal_subscribe(connection, "org.freedesktop.portal.Desktop", "org.freedesktop.portal.Request",
+        "Response", requestPath.ascii().data(), nullptr, G_DBUS_SIGNAL_FLAGS_NO_MATCH_RULE,
+        [](GDBusConnection* connection, const char*, const char*, const char*, const char*, GVariant* parameters, gpointer userData) {
+            GRefPtr<WebKitPrintOperation> printOperation(static_cast<WebKitPrintOperation*>(userData));
+            auto* priv = printOperation->priv;
+            GRefPtr<GVariant> options;
+            uint32_t response;
+
+            RELEASE_ASSERT(priv->signalId);
+            g_dbus_connection_signal_unsubscribe(connection, priv->signalId);
+            priv->signalId = 0;
+
+            g_variant_get(parameters, "(u@a{sv})", &response, &options.outPtr());
+
+            if (response) {
+                priv->preparePrintResponse = PreparePrintResponse(WEBKIT_PRINT_OPERATION_RESPONSE_CANCEL);
+                return;
+            }
+
+            GRefPtr<GVariant> printSettingsVariant(g_variant_lookup_value(options.get(), "settings", G_VARIANT_TYPE_VARDICT));
+            GRefPtr<GVariant> pageSetupVariant(g_variant_lookup_value(options.get(), "page-setup", G_VARIANT_TYPE_VARDICT));
+
+            priv->printSettings = gtk_print_settings_new_from_gvariant(printSettingsVariant.get());
+            priv->pageSetup = gtk_page_setup_new_from_gvariant(pageSetupVariant.get());
+
+            uint32_t token;
+            g_variant_lookup(options.get(), "token", "u", &token);
+
+            priv->preparePrintResponse = PreparePrintResponse(token);
+        }, g_object_ref(printOperation), g_object_unref);
+
+    GVariantBuilder options;
+    g_variant_builder_init(&options, G_VARIANT_TYPE_VARDICT);
+    g_variant_builder_add(&options, "{sv}", "handle_token", g_variant_new_string(token.ascii().data()));
+
+    const char* title = _("Print Web Page");
+    GRefPtr<GtkPageSetup> pageSetup = priv->pageSetup ? priv->pageSetup : adoptGRef(gtk_page_setup_new());
+    GRefPtr<GtkPrintSettings> printSettings(priv->printSettings ? gtk_print_settings_copy(priv->printSettings.get()) : gtk_print_settings_new());
+
+    GRefPtr<GVariant> printSettingsVariant(gtk_print_settings_to_gvariant(printSettings.get()));
+    GRefPtr<GVariant> pageSetupVariant(gtk_page_setup_to_gvariant(pageSetup.get()));
+    GRefPtr<GVariant> arguments(g_variant_new("(ss@a{sv}@a{sv}a{sv})", "", title, printSettingsVariant.get(), pageSetupVariant.get(), &options));
```

Comment:
Since `gtk_print_settings_to_gvariant` and `gtk_page_setup_to_gvariant` return a floating ref, we can just pass it directly to `g_variant_new` instead of converting tu full ref and then passing to `g_variant_new`